pkg/operator/ceph/cluster/osd/provision_spec.go

```diff
@@ -157,17 +157,31 @@ func (c *Cluster) provisionPodTemplateSpec(osdProps osdProperties, restart v1.Re
 		podSpec.DNSPolicy = v1.DNSClusterFirstWithHostNet
 	}
 	if osdProps.onPVC() {
-		// The "all" placement is applied separately so it will have lower priority.
-		// We want placement from the storageClassDeviceSet to be applied and override
-		// the "all" placement if there are any overlapping placement settings.
-		c.spec.Placement.All().ApplyToPodSpec(&podSpec)
-		// Apply storageClassDeviceSet PreparePlacement
-		// If nodeAffinity is specified both in the device set and "all" placement,
-		// they will be merged.
+		// when "onlyApplySpecificOSDPlacement" false, in case of NodeAffinity it will merge both spec.placement.All() and storageClassDeviceSets.preparePlacement,
+		// in case of other preparePlacement rule like PodAffinity, PodAntiAffinity... it will override last prepareplacement with the current preparePlacement applied,
+		// technically in another word for prepareplacement apart from NodeAffinity, it will same as "onlyApplySpecificOSDPlacement" true as storageClassDeviceSets.preparePlacement
+		// will override spec.placement.All(). See ApplyToPodSpec().
+
+		// when "onlyApplySpecificOSDPlacement" true, it will only apply storageClassDeviceSets.preparePlacement
+		if !c.spec.Storage.OnlyApplySpecificOSDPlacement {
+			// apply spec.placement.All()
+			c.spec.Placement.All().ApplyToPodSpec(&podSpec)
+		}
+		// apply storageClassDeviceSets.preparePlacement
 		osdProps.getPreparePlacement().ApplyToPodSpec(&podSpec)
```

Comment:
How about factoring out lines 160-171 to a helper function? It looks like the only difference between the four places this is executed is the placement on the last line. For example, `osdProps.getPreparePlacement()`, or `c.spec.Placement[cephv1.KeyOSDPrepare]` would become a param to that function.

Reply:
done.